crates/polars-pipe/src/operators/chunks.rs

```diff
@@ -28,5 +28,152 @@ impl DataChunk {
 }
 
 pub(crate) fn chunks_to_df_unchecked(chunks: Vec<DataChunk>) -> DataFrame {
-    accumulate_dataframes_vertical_unchecked(chunks.into_iter().map(|c| c.data))
+    let mut combiner = SemicontiguousVstacker::<4194304>::new();
+    let mut frames_iterator = chunks
+        .into_iter()
+        .flat_map(|c| combiner.add(c.data))
+        .peekable();
+    if frames_iterator.peek().is_some() {
+        let mut result = accumulate_dataframes_vertical_unchecked(frames_iterator);
+        if let Some(df) = combiner.finish() {
+            let _ = result.vstack_mut(&df);
+        }
+        result
+    } else {
+        // The presumption is that this function is never called with empty
+        // data, cause that'll cause accumulate_dataframes_vertical_unchecked to
+        // error, so if we haven't gotten any data we can safely assume it's in
+        // the combiner buffer.
+        combiner.finish().unwrap()
+    }
+}
+
+/// Combine `DataFrame`s into sufficiently large contiguous memory allocations
+/// that we don't suffer from overhead of many small writes. At the same time,
+/// don't use too much memory by creating overly large allocations.
+///
+/// The assumption is that added `DataFrame`s are already in the correct order,
+/// and can therefore be combined.
+///
+/// 4 MB was chosen based on some empirical experiments that showed it to
+/// be decently faster than lower or higher values, and it's small enough
+/// it won't impact memory usage significantly.
+#[derive(Clone)]
+pub(crate) struct SemicontiguousVstacker<const BIG_ENOUGH: usize = 4194304> {
```

Comment:
I don't think we have to make this a const parameter. That would require a recompile every instansiation that's not really worth it perf-wise.

Reply:
It's just there for testing to save memory, but it's not necessary for testing, so I can remove.